store/iavlstore.go

```diff
@@ -0,0 +1,253 @@
+package store
+
+import (
+	"path"
+	"path/filepath"
+	"strings"
+
+	"github.com/pkg/errors"
+
+	"github.com/tendermint/iavl"
+	dbm "github.com/tendermint/tmlibs/db"
+)
+
+// NewIAVLLoader returns a CommitterLoader that returns
+// an IAVLCommitter
+func NewIAVLLoader(dbName string, cacheSize int, history uint64) CommitterLoader {
+	l := iavlLoader{
+		dbName:    dbName,
+		cacheSize: cacheSize,
+		history:   history,
+	}
+	return CommitterLoader(l.Load)
+}
+
+// IAVLCommitter Implements IterKVStore and Committer
+type IAVLCommitter struct {
+	// we must store the last height here, as it is needed
+	// for saving the versioned tree
+	lastHeight uint64
+
+	// history is how many old versions we hold onto,
+	// uses a naive "hold last X versions" algorithm
+	history uint64
+
+	// this is all historical data and connection to
+	// the db
+	tree *iavl.VersionedTree
+
+	// this is the current working state to be saved
+	// on the next commit
+	IAVLStore
+}
+
+// NewIAVLCommitter properly initializes a committer
+// that is ready to use as a IterKVStore
+func NewIAVLCommitter(tree *iavl.VersionedTree,
+	lastHeight uint64, history uint64) *IAVLCommitter {
+	i := &IAVLCommitter{
+		tree:       tree,
+		lastHeight: lastHeight,
+		history:    history,
+	}
+	i.updateStore()
+	return i
+}
+
+// Commit syncs the working state and
+// saves another version to the db
+func (i *IAVLCommitter) Commit() CommitID {
+	// TODO: sync working state??
+	// I think this is done already just by writing to tree.Tree()
+
+	// save a new version
+	i.lastHeight++
+	hash, err := i.tree.SaveVersion(i.lastHeight)
+	if err != nil {
+		// TODO: do we want to extend Commit to
+		// allow returning errors?
+		panic(err)
+	}
+
+	// now point working state to the new status
+	i.updateStore()
+
+	// release an old version of history
+	if i.history <= i.lastHeight {
+		release := i.lastHeight - i.history
+		i.tree.DeleteVersion(release)
+	}
+
+	return CommitID{
+		Version: i.lastHeight,
+		Hash:    hash,
+	}
+}
+
+// store returns a wrapper around the current writable state
+func (i *IAVLCommitter) updateStore() {
+	i.IAVLStore = IAVLStore{i.tree.Tree()}
+}
+
+var _ CacheWrappable = (*IAVLCommitter)(nil)
```

Comment:
lets move these above the struct def so we know right away what it implements (the comments say it implements these things, might as well put these assertions close by!)